From e9e2072e731d403c41bfe5403bd28d506889c906 Mon Sep 17 00:00:00 2001
From: Ben Sartor <ben@simlar.org>
Date: Thu, 15 Oct 2015 04:04:47 +0200
Subject: [PATCH 2/2] fixed crash when calling linphone_call_get_remote_params
 in state LinphoneCallOutgoingInit

---
 coreapi/call_params.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/coreapi/call_params.c b/coreapi/call_params.c
index 63dc25c..8b900c6 100644
--- a/coreapi/call_params.c
+++ b/coreapi/call_params.c
@@ -73,6 +73,9 @@ SalStreamDir get_video_dir_from_call_params(const LinphoneCallParams *params) {
 }
 
 void linphone_call_params_set_custom_headers(LinphoneCallParams *params, const SalCustomHeader *ch){
+	if (params == NULL){
+		return;
+	}
 	if (params->custom_headers){
 		sal_custom_header_free(params->custom_headers);
 		params->custom_headers = NULL;
-- 
2.6.1

