From 2ec8bb88b4f63344adb34e1c2f77da3a32b7f5ff Mon Sep 17 00:00:00 2001
From: Ben Sartor <mail@ben-sartor.de>
Date: Thu, 21 Aug 2014 21:16:41 +0200
Subject: [PATCH] ios sound now stops shutdown timer before destroying
 AudioUnit

---
 src/audiofilters/msiounit.m | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/audiofilters/msiounit.m b/src/audiofilters/msiounit.m
index 27cfefe..ea7e332 100644
--- a/src/audiofilters/msiounit.m
+++ b/src/audiofilters/msiounit.m
@@ -168,6 +168,7 @@ struct au_filter_write_data{
 };
 
 static void  stop_audio_unit (au_card_t* d);
+static void  check_audio_unit (au_card_t* d);
 
 
 /*
@@ -254,6 +255,7 @@ static void au_init(MSSndCard *card){
 
 static void au_uninit(MSSndCard *card){
 	au_card_t *d=(au_card_t*)card->data;
+	check_audio_unit(d);
 	stop_audio_unit(d);
 	ms_mutex_destroy(&d->mutex);
 	ms_free(d);
-- 
2.0.1

